crates/ra_hir_ty/src/expr.rs

```diff
@@ -310,3 +315,87 @@ pub fn record_pattern_missing_fields(
     }
     Some((variant_def, missed_fields, exhaustive))
 }
+
+pub fn unsafe_expressions(
+    db: &dyn HirDatabase,
+    infer: &InferenceResult,
+    def: DefWithBodyId,
+) -> Vec<ExprId> {
+    let mut unsafe_expr_ids = vec![];
+    let body = db.body(def);
+    for (id, expr) in body.exprs.iter() {
+        match expr {
+            Expr::Call { callee, .. } => {
+                if infer
+                    .method_resolution(/* id */ *callee)
```

Comment:
Can't seem to get this section to work. How do I get the `FunctionData` once I have an `Expr::Call` or `Expr::MethodCall`?

Reply:
Look at the inferred type of `callee`. There's various callable types, but most of them can't be unsafe I think; for normal functions, the type will be `Ty::Apply(TypeCtor::FnDef(CallableDef::FunctionId(function_id)), ..)`. Fn pointers can also be unsafe, I think, but we don't track that yet.

Reply:
For a method call, the code below should be correct.

Reply:
So I've been debugging this morning and the method call code doesn't work. `infer.method_resolution(id)` always returns `None`. Am I possibly setting up the inference incorrectly?